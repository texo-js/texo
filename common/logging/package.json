{
  "name": "@texo/logging",
  "version": "0.0.0",
  "description": "Logging extensions for namespaced logging and filtering.",
  "author": "Texo Authors <texo@defize.co.uk>",
  "homepage": "https://github.com/compile-this/texo/tree/master/api/gateway#readme",
  "license": "MIT",
  "type": "module",
  "main": "./lib/index.js",
  "module": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "exports": {
    "import": "./lib/index.js"
  },
  "directories": {
    "lib": "lib"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/compile-this/texo.git"
  },
  "bugs": {
    "url": "https://github.com/compile-this/texo/issues"
  },
  "devDependencies": {
    "@types/chalk": "^2.2.0",
    "@types/figlet": "^1.2.0",
    "@types/triple-beam": "^1.3.1"
  },
  "dependencies": {
    "logform": "^2.2.0",
    "triple-beam": "^1.3.0",
    "winston": "^3.2.1",
    "winston-transport": "^4.3.0"
  }
}
