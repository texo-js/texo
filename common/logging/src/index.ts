export * from './logging';
export * from './filters';
export * from './namespace-set';
 
import winston, { Logger } from 'winston';

export { Logger };
export const { format, transports } = winston;
