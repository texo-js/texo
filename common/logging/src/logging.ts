import winston, { Logger, LoggerOptions } from 'winston';

const { createLogger: createWinstonLogger, transports, format } = winston;

interface Metadata extends Record<string, any> {
  ns?: string;
}

interface LoggerEx extends Logger {
  defaultMeta: Metadata | undefined;
}

namespace Loggers {
  export function create({ options = {}, namespace = '' } : { options?: LoggerOptions, namespace?: string } = {}) : Logger {
    options.defaultMeta = { ns: namespace, ...(options.defaultMeta || {}) };
    const logger = createWinstonLogger(options);
    
    return logger;
  }

  export function createChild({ parent, namespace } : { parent: Logger, namespace: string }) {
    const logger = parent.child({}) as LoggerEx;

    const metadata = logger.defaultMeta || {};
    const namespaceRoot = metadata.ns || '';
    const ns = `${namespaceRoot}/${namespace}`;
    
    logger.defaultMeta = { ...metadata, ns };

    return logger;
  }

  let systemLogger = createNullLogger();

  export function setSystem(logger: Logger) {
    systemLogger = logger;
  }

  export function getSystem() {
    return systemLogger;
  }

  function createNullLogger(): Logger {
    return Loggers.create({ options: { level: 'debug', transports: [] } });
  }
}

export { Loggers, LoggerOptions, transports, format }