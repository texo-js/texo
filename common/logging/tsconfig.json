{
  "extends": "../../tsconfig.base.json",

  "compilerOptions": {
    "composite": true,
    "rootDir": "./src",
    "outDir": "./lib",
    "typeRoots": [ "../../@types/", "./node_modules/@types/" ]
  },

  "include": [
    "./src/**/*"
  ],

  "exclude": [
    "./src/**/*.test.ts"
  ]
}