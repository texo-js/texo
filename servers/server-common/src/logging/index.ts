export * from './apollo';
export * from './logger';