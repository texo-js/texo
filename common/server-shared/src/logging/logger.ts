import { NamespacedLogger } from '@texo/logging';

let _logger: NamespacedLogger | null = null;

function setLogger(logger: NamespacedLogger) {
  _logger = logger;
}

export { _logger as logger, setLogger };
