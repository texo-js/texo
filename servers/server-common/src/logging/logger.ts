import { EOL } from 'os';

import { Logger, Loggers, transports, format, Filters, Format, NamespaceSet } from '@texo/logging';

namespace Defaults {
  let _defaultLogger: Logger;

  export const DefaultNamespaceSet = new NamespaceSet();

  export const DefaultConsoleFormat: Format = format.combine(
    Filters.create({ filterLevel: 'info', namespaces: DefaultNamespaceSet }),
    format.timestamp(),
    format.colorize({ level: true, all: false, message: false }),
    format.metadata({ fillExcept: [ 'level', 'ns', 'timestamp', 'message' ] }),
    format.printf(info => `${info.timestamp} ${info.level} [${info.ns}] ${info.message}${EOL}${JSON.stringify(info.metadata)}${EOL}`)
  );

  export function getDefaultLogger() : Logger {
    if (!_defaultLogger) {
      _defaultLogger = createDefaultLogger();
    }
  
    return _defaultLogger;
  }

  function createDefaultLogger() : Logger {
    return Loggers.create({ options: { level: 'debug', format: DefaultConsoleFormat, transports: transports.Console  } })
  }
}

let _systemLogger: Logger;

function getSystemLogger() : Logger {
  if (!_systemLogger) {
    _systemLogger = Defaults.getDefaultLogger();
  }

  return _systemLogger;
}

function setSystemLogger(logger: Logger) {
  _systemLogger = logger;
}

export {
  Logger,
  Loggers,
  Filters,
  NamespaceSet,
  Defaults,
  format as Formats,
  transports as Transports,
  getSystemLogger,
  setSystemLogger
};
