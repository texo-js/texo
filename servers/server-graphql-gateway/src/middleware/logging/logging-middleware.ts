import Koa from 'koa';
import { Loggers } from '@texo/logging';

export interface LoggingMiddleware {
  logging(): Koa.Middleware
}

export class DefaultLoggingMiddleware implements LoggingMiddleware {
  logging(): Koa.Middleware {
    const logger = Loggers.namespace('texo/koa');

    return async (ctx: Koa.Context, next: Koa.Next) => {
      const metadata = {
        correlationId: ctx.state.correlationId,
        request: {
          method: ctx.request.method,
          url: ctx.request.url
        }
      };

      logger.info('http request started', metadata);

      try {
        await next();
      } catch (e) {
        logger.error(e, metadata);
      } finally {
        logger.info('http request complete', metadata);
      }
    }
  }
}