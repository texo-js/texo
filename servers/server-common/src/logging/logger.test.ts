import { logger, setLogger } from './logger';
import { createLogger, NamespaceFilters } from '@texo/logging'

describe('the default logger can be managed', function () {
  test('the logger can be set and retrieved', function () {
      const filters = new NamespaceFilters('debug', '*');
      const theLogger = createLogger('test', filters, {});
    
      setLogger(theLogger);

      expect(logger).toBe(theLogger);
  });
});