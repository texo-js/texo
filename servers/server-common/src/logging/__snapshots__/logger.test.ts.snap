// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger The default console format rendering correctly formats the log output 1`] = `
"2020-07-02T16:00:04.102Z [32minfo[39m [/level-2/level-2] hello
{\\"something\\":\\"else\\"}
"
`;
